l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol

```diff
@@ -0,0 +1,416 @@
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Disable the initialization to prevent Parity hack.
+    /// @param _wethToken Address of WETH on deployed chain
+    /// @param _baseTokenAddress Address of Base token
+    constructor(address _wethToken, address _assetRouter, address _baseTokenAddress) {
+        _disableInitializers();
+        ASSET_ROUTER = IAssetRouterBase(_assetRouter);
+        WETH_TOKEN = _wethToken;
+        BASE_TOKEN_ADDRESS = _baseTokenAddress;
+    }
+
+    /// @notice Sets token beacon used by bridged ERC20 tokens deployed by NTV.
+    /// @dev we don't call this in the constructor, as we need to provide factory deps
+    function setBridgedTokenBeacon() external {
+        if (address(bridgedTokenBeacon) != address(0)) {
+            revert AddressMismatch(address(bridgedTokenBeacon), address(0));
+        }
+        address bridgedStandardToken = address(new BridgedStandardERC20{salt: bytes32(0)}());
+        bridgedTokenBeacon = new UpgradeableBeacon{salt: bytes32(0)}(bridgedStandardToken);
+        bridgedTokenBeacon.transferOwnership(owner());
+    }
+
+    /// @notice Registers tokens within the NTV.
+    /// @dev The goal was to allow bridging native tokens automatically, by registering them on the fly.
+    /// @notice Allows the bridge to register a token address for the vault.
+    /// @notice No access control is ok, since the bridging of tokens should be permissionless. This requires permissionless registration.
+    function registerToken(address _nativeToken) external {
+        if (_nativeToken == WETH_TOKEN) {
+            revert TokenNotSupported(WETH_TOKEN);
+        }
+        require(_nativeToken == BASE_TOKEN_ADDRESS || _nativeToken.code.length > 0, "NTV: empty token");
+        bytes32 assetId = DataEncoding.encodeNTVAssetId(block.chainid, _nativeToken);
+        ASSET_ROUTER.setAssetHandlerAddressThisChain(bytes32(uint256(uint160(_nativeToken))), address(this));
+        tokenAddress[assetId] = _nativeToken;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            FINISH TRANSACTION FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    ///  @inheritdoc IAssetHandler
+    /// @notice Used when the chain receives a transfer from L1 Shared Bridge and correspondingly mints the asset.
+    /// @param _chainId The chainId that the message is from.
+    /// @param _assetId The assetId of the asset being bridged.
+    /// @param _data The abi.encoded transfer data.
+    function bridgeMint(
+        uint256 _chainId,
+        bytes32 _assetId,
+        bytes calldata _data
+    ) external payable override onlyAssetRouter whenNotPaused {
+        // Either it was locked before, therefore is not zero, or it is sent from remote chain and standard erc20 will be deployed
+        address token = tokenAddress[_assetId];
+        if (chainBalance[_chainId][token] > 0) {
```

Comment:
at the start of the migration to NTV some of chainBalances will be zero (until those are manually supplied for each pair of chain/token), does this part take it into account?

Reply:
If the chainBalance is zero it will try to deploy the token, and it will do the assetIdCheck. There we check that the assetId was registered on the L2, so the deployment will fail.